compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/optimizer/BIRRecordValueOptimizer.java

```diff
@@ -192,7 +192,8 @@ private boolean containsOnlyConstantLoad(BIRNode.BIRFunction defaultFunction) {
         }
         BIRNode.BIRBasicBlock firstBB = defaultFunction.basicBlocks.get(0);
         BIRNode.BIRBasicBlock secondBB = defaultFunction.basicBlocks.get(1);
-        if (!secondBB.instructions.isEmpty() || secondBB.terminator.kind != InstructionKind.RETURN) {
+        if (!secondBB.instructions.isEmpty() || secondBB.terminator.kind != InstructionKind.RETURN ||
```

Comment:
The cause of the issue is that we rely on only the `ConstantLoad` instructions and the `RETURN` terminator to determine the default value function. This could lead to identifying other possible cases as the default value function. 
But this fix will only eliminate one case of it, that the `firstBB` terminator can be a function call. 
I would suggest having a logic that is more specific to default functions. Something similar to this?
```java
firstBB.instructions.get(0).lhsOp.variableDcl
        .kind == VarKind.RETURN // to specify the return variab;e is already set
```
or 
```java
firstBB.terminator.kind != InstructionKind.GOTO
```
@rdulmina WDYT?

Reply:
+1 for the first apporach. We can modify the switch case as below
```
        return switch (firstBB.instructions.size()) {
            case 1 -> firstBB.instructions.get(0).kind == CONST_LOAD && firstBB.instructions.get(0).lhsOp.variableDcl
                    .kind == VarKind.RETURN;
            case 2 -> firstBB.instructions.get(0).kind == CONST_LOAD && firstBB.instructions.get(1).kind == TYPE_CAST
                    && firstBB.instructions.get(1).lhsOp.variableDcl.kind == VarKind.RETURN;
            default -> false;
        };
 ```

Reply:
You are correct. The current fix stops the optimization from happening to other functions as well. If I understood correctly, the optimization is trying to drop the generated default function for cases such as following,
```ballerina
type Quz record {
    string c = "hello";
};
```
And directly load the constant when the record value is created. 

The problem seems to be only with default functions with arguments. Please correct me if I'm wrong. I will check the suggestions you gave.

Reply:
@rdulmina your suggestion worked. Added the change to 017c4bcfacc0ee63616acd47cfe89e610b32b8ea.